src/index.ts

```diff
@@ -1,155 +1,61 @@
-      .update(JSON.stringify(body))
-      .digest('hex')
-    const priorHash = (
-      await readFile('.command-hash').catch(() => '')
-    ).toString()
-    if (commandHash === priorHash) {
-      console.log('✅ Commands are up-to-date')
-      return
-    }
-
-    console.time('updating commands')
-    await rest.put(
-      Routes.applicationGuildCommands(
-        process.env.CLIENT_ID,
-        process.env.GUILD_ID
-      ),
-      {
-        body,
-      }
-    )
-
-    writeFile('.command-hash', commandHash)
-
-    console.timeEnd('updating commands')
-  } catch (error) {
-    console.log(error)
-  }
-}
-
-let client: Discord.Client
-
-async function main() {
-  updateCommands()
-
-  console.time('discord client ready')
-  client = new Discord.Client({
-    intents: [
-      Intents.FLAGS.GUILDS,
-      Intents.FLAGS.GUILD_MESSAGES,
-      Intents.FLAGS.GUILD_MESSAGE_REACTIONS,
-      Intents.FLAGS.GUILD_EMOJIS_AND_STICKERS,
-    ],
-  })
-
-  // fixes `TypeError: Do not know how to serialize a BigInt` for store.dispatch(commandInteraction(interaction));
-  // https://github.com/GoogleChromeLabs/jsbi/issues/30#issuecomment-1006086291
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  ;(BigInt.prototype as any).toJSON = function () {
-    return this.toString()
-  }
-
-  client.on('interactionCreate', async (interaction) => {
-    if (!interaction.isCommand()) return
-    store.dispatch(commandUsed(interaction))
-    console.log(`command ${interaction.commandName}`)
-    console.time(interaction.commandName)
-    try {
-      await interaction.deferReply()
-      const command = commands.get(interaction.commandName)
-      if (!command) {
-        interaction.editReply(`Command not found ${command}`)
-        return
-      }
-      await command.execute(interaction)
-    } catch (e) {
-      console.error(e)
-      await interaction.followUp(
-        `Command \`${interaction.commandName}\` failed with error: \`${e}\``
+  assertEnv,
+  createClient,
+  gameClock,
+  waitForWinner,
+} from '@adventure-bot/game'
+import { ReduxState } from '@adventure-bot/store'
+
+let client: Client
+
+const startGame: (gameOptions: {
+  clientId: string
+  token: string
+  channelId: string
+}) => ReturnType<typeof createClient> = async ({
+  clientId,
+  token,
+  channelId,
+}) => {
+  let lastState: ReduxState
+
+  return createClient({
+    type: 'discord',
+    token,
+    clientId,
+    channelId,
+    onBeforeInteraction: async (gameState) => {
+      // Keep track of what the game state is prior to the users interaction
+      lastState = gameState
+    },
+    onAfterInteraction: async (gameState) => {
+      const didStateChange = Boolean(
+        Object.keys(diff(lastState, gameState)).length
       )
-    }
-    console.timeEnd(interaction.commandName)
-  })
-  client.on('error', (e) => {
-    console.error('Discord client error!', e)
-  })
+      if (didStateChange) {
+        // Persist to DB
+      }
```

Comment:
Isn't this essentially what redux-persist does?

Reply:
Yes, I'm going to change this after we'd spoken about the site design/use